File: frontend/src/api/media.ts
import { buildApiUrl, deleteRequest, getJson, patchJson, postForm } from './http'
import { clampQueryLimit } from './limits'
import type { ListResponse, Media, MediaListItem } from '../types/models'

export function listMedia(limit = 50, offset = 0, includeFilePath = true): Promise<ListResponse<MediaListItem>> {
  return getJson<ListResponse<MediaListItem>>(
    `/media?limit=${clampQueryLimit(limit)}&offset=${offset}&include_file_path=${String(includeFilePath)}`,
  )
}

export function queryMedia(params: {
  nameContains?: string
  speciesIds?: number[]
  tagId?: number
  titleContains?: string
  mimeType?: string
  plantId?: number
  minSize?: number
  maxSize?: number
  limit?: number
  offset?: number
  includeFilePath?: boolean
}): Promise<Media[]> {
  const query = new URLSearchParams()
  if (params.nameContains) query.set('name_contains', params.nameContains)
  if (params.speciesIds && params.speciesIds.length > 0) {
    for (const speciesId of params.speciesIds) {
      query.append('species_ids', String(speciesId))
    }
  }
  if (params.tagId !== undefined) query.set('tag_id', String(params.tagId))
  if (params.titleContains) query.set('title_contains', params.titleContains)
  if (params.mimeType) query.set('mime_type', params.mimeType)
  if (params.plantId !== undefined) query.set('plant_id', String(params.plantId))
  if (params.minSize !== undefined) query.set('min_size', String(params.minSize))
  if (params.maxSize !== undefined) query.set('max_size', String(params.maxSize))
  query.set('limit', String(clampQueryLimit(params.limit)))
  query.set('offset', String(params.offset ?? 0))
  query.set('include_file_path', String(params.includeFilePath ?? true))

  return getJson<Media[]>(`/media/query?${query.toString()}`)
}

export function getMediaById(id: number, includeFilePath = true): Promise<Media> {
  return getJson<Media>(`/media/${id}?include_file_path=${String(includeFilePath)}`)
}

export function uploadMedia(file: File, title?: string, plantId?: number, tagId?: number): Promise<Media> {
  const form = new FormData()
  form.append('file', file)
  if (title) form.append('title', title)
  if (plantId !== undefined) form.append('plant_id', String(plantId))
  if (tagId !== undefined) form.append('tag_id', String(tagId))
  return postForm<Media>('/media', form)
}

export function updateMedia(id: number, payload: { title?: string | null; mime_type?: string | null; size?: number | null; plant_id?: number | null; tag_id?: number | null }): Promise<Media> {
  return patchJson<Media, { title?: string | null; mime_type?: string | null; size?: number | null; plant_id?: number | null; tag_id?: number | null }>(`/media/${id}`, payload)
}

export function deleteMedia(id: number): Promise<void> {
  return deleteRequest(`/media/${id}`)
}

export function mediaFileUrl(id: number): string {
  return buildApiUrl(`/media/${id}/file`)
}

File: src/api/routers/media.py
"""Media API routes."""

from pathlib import Path

from fastapi import APIRouter, File, Form, HTTPException, Query, UploadFile, status
from fastapi.responses import FileResponse

from src.api.schemas import MediaUpdate
from src.media.image import delete_image, retrieve_image, save_image
from src.models.media import Media, MediaCreate, MediaListResponse
from src.services.media_service import (
    create_media,
    delete_media_by_filename_key,
    delete_media_by_id,
    get_media_by_id,
    list_media,
    query_media,
    update_media,
)
from src.services.plant_service import get_plant_by_id as get_plant_record_by_id
from src.services.species_service import get_species_by_id as get_species_record_by_id
from src.services.tag_service import get_tag_by_id as get_tag_record_by_id
from src.settings import settings

router = APIRouter(prefix="/media", tags=["media"])


@router.get("", response_model=MediaListResponse)
def list_media_endpoint(
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    include_file_path: bool = False,
) -> MediaListResponse:
    return list_media(limit=limit, offset=offset, include_file_path=include_file_path)


@router.get("/query", response_model=list[Media])
def query_media_endpoint(
    name_contains: str | None = None,
    title_contains: str | None = None,
    mime_type: str | None = None,
    plant_id: int | None = None,
    species_ids: list[int] | None = Query(default=None),
    tag_id: int | None = None,
    min_size: int | None = Query(default=None, ge=0),
    max_size: int | None = Query(default=None, ge=0),
    limit: int = Query(default=50, ge=1, le=200),
    offset: int = Query(default=0, ge=0),
    include_file_path: bool = False,
) -> list[Media]:
    return query_media(
        name_contains=name_contains,
        title_contains=title_contains,
        mime_type=mime_type,
        plant_id=plant_id,
        species_ids=species_ids,
        tag_id=tag_id,
        min_size=min_size,
        max_size=max_size,
        limit=limit,
        offset=offset,
        include_file_path=include_file_path,
    )


@router.get("/{media_id}", response_model=Media)
def get_media_endpoint(media_id: int, include_file_path: bool = False) -> Media:
    record = get_media_by_id(media_id, include_file_path=include_file_path)
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")
    return record


@router.get("/{media_id}/file", response_class=FileResponse)
def get_media_file_endpoint(media_id: int) -> FileResponse:
    record = get_media_by_id(media_id)
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")
    return retrieve_image(record.filename)


@router.post("", response_model=Media, status_code=status.HTTP_201_CREATED)
async def upload_media_endpoint(
    file: UploadFile = File(...),
    title: str | None = Form(default=None),
    plant_id: int | None = Form(default=None),
    tag_id: int | None = Form(default=None),
    species_id: int | None = Form(default=None),
) -> Media:
    if plant_id is not None and not get_plant_record_by_id(plant_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plant not found")
    if tag_id is not None and not get_tag_record_by_id(tag_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    if species_id is not None and not get_species_record_by_id(species_id):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Species not found")

    unique_filename = await save_image(file)
    file_path = Path(settings.media_path) / unique_filename

    payload = MediaCreate(
        filename=unique_filename,
        title=title,
        mime_type=file.content_type or "application/octet-stream",
        size=file_path.stat().st_size,
        plant_id=plant_id,
        tag_id=tag_id,
        species_id=species_id,
    )
    return create_media(payload)


@router.patch("/{media_id}", response_model=Media)
def update_media_endpoint(media_id: int, payload: MediaUpdate) -> Media:
    kwargs: dict = {}
    if "title" in payload.model_fields_set:
        kwargs["title"] = payload.title
    if "mime_type" in payload.model_fields_set:
        kwargs["mime_type"] = payload.mime_type
    if "size" in payload.model_fields_set:
        kwargs["size"] = payload.size
    if "plant_id" in payload.model_fields_set:
        kwargs["plant_id"] = payload.plant_id
    if "tag_id" in payload.model_fields_set:
        kwargs["tag_id"] = payload.tag_id
    if "species_id" in payload.model_fields_set:
        kwargs["species_id"] = payload.species_id

    if "plant_id" in kwargs and kwargs["plant_id"] is not None and not get_plant_record_by_id(kwargs["plant_id"]):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Plant not found")
    if "tag_id" in kwargs and kwargs["tag_id"] is not None and not get_tag_record_by_id(kwargs["tag_id"]):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Tag not found")
    if "species_id" in kwargs and kwargs["species_id"] is not None and not get_species_record_by_id(kwargs["species_id"]):
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Species not found")

    updated = update_media(media_id, **kwargs)
    if not updated:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")
    return updated


@router.delete("/{media_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_media_endpoint(media_id: int) -> None:
    record = get_media_by_id(media_id)
    if not record:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")

    delete_image(record.filename)
    deleted = delete_media_by_id(media_id)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")


@router.delete("/by-filename/{filename}", status_code=status.HTTP_204_NO_CONTENT)
def delete_media_by_filename_endpoint(filename: str) -> None:
    delete_image(filename)
    deleted = delete_media_by_filename_key(filename)
    if not deleted:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Media not found")
